### PR TITLE
stateless crypto

### PR DESCRIPTION
**Pull request by dominictarr**
This change removes the `state.js` module and replaces it with `stateless.js`.
The objective here is to have all crypto generated by pure functions, this makes it easy to have a set of test vectors, which makes it easy to port.

There are a few crypto functions which apply to both the client and server. (`initialize`, `createChallenge`, `clean`, and `verifyChallenge`) the rest are divided into client and server. All functions that with "create" in the name return a buffer, but do not change state. All functions with "verify" take the state and a buffer (which was the output of the remote's previous create function) and mutate the state object and return it. mutating the state makes it not a "pure" pure function, but we don't want to have all our secrets just floating around, so this means we might have a chance to zero them, which reduces exposure to say, unsafe memory access.

It's getting late here, but the next step is to generate a set of test vectors. They are really simple, just an array of `[methodname, [args...], returnValue]`
This would be most portable in JSON format, with base64 or hex encoded buffers, I figure this should be easy on most languages?

Then, an implementer just needs to implement the functions, and iterate over the test vector and make sure every output matches. The logic of the handshake is pretty straightforward after that.

@cryptix @keks @david415 @Kodest

### CODE REVIEW

**Comment by cryptix** [comment]
welcome this a lot!

ps: just a small typo here:

> and return a mutate the state object and return it.

 I guess _and return a mutated version of the state object_?

**Comment by dominictarr** [comment]
@cryptix thanks

**Comment by dominictarr** [comment]
okay, the protocol logic and the crypto are now clearly separated. next I'm gonna generate test vectors by wrapping the crypto functions in loggers, write that to a json file, and then read that back as test vectors.

**Comment by dominictarr** [comment]
still a little bit of tidying up due before this is mergeable, but I got test vectors
https://github.com/auditdrivencrypto/test-secret-handshake

that matches the reference implementation (though, only on packet type differs) [not the spec](https://github.com/auditdrivencrypto/secret-handshake/issues/7)

**Comment by dominictarr** [comment]
@david415

**Comment by Kodest** [comment]
glad to see those test vectors, they will be very useful for my planned C++ implementation. (I think json format is approachable.)

File: crypto.js [hunk 1]
```diff
@@ -18,52 +17,50 @@ var concat = Buffer.concat
...
+//both client and server
 
-function State (app_key, local, remote, seed) {
+exports.initialize = function (state) {
 
-  if(!(this instanceof State)) return new State(app_key, local, remote, seed)
+  if(state.seed) state.local = from_seed(state.seed)
 
-  if(seed) local = from_seed(seed)
+  //TODO: sodium is missing box_seed_keypair. should make PR for that.
```
Comment (emilbayes) [inline crypto.js]: You could always switch to @mafintosh `sodium-native` ^^

**Comment by david415** [comment]
looks good.

**Comment by dominictarr** [comment]
I think I might change from base64 to hex - hex is what you'd see if you use tools like hexdump, and if you really need to, it is easier to implement hex than base64, also it's less ambigious whether a normal string is hex or not, where as most printable ascii strings are base64

**Comment by david415** [comment]
ok yeah hex is fine

File: crypto.js (hunk 1 above)
Comment (dominictarr) [inline crypto.js]: yup, this is a good suggestion.

**Comment by dominictarr** [comment]
merged into 1.1.3,
test vectors available at https://github.com/auditdrivencrypto/test-secret-handshake

would implementers like to move their implementations into a single org? this is probably easier for users, and I feel gives the impression that yes, a few sets of eyes have gone over the protocol & implementations.

File: crypto.js [hunk 2]
```diff
@@ -18,52 +17,50 @@ var concat = Buffer.concat
...
+    kx_sk: kx_sk,
+    publicKey: state.local.publicKey,
+    secretKey: state.local.secretKey,
+    app_mac: auth(kx_pk, state.app_key)
+  }
 
-var proto = State.prototype
+  state.local.kx_pk = kx_pk
+  state.local.kx_sk = kx_sk
+  state.local.app_mac = auth(kx_pk, state.app_key)
```
Comment (Kodest) [inline crypto.js]: I am not very good at javascript but aren't these assignments redundant? `state.local` has been initialized several lines above: `kx_pk`, `kx_sk` and `app_mac` already have those values.

Comment (dominictarr) [inline crypto.js]: that is correct, good spotting.

Comment (dominictarr) [inline crypto.js]: fixed in 1.1.7

**Pull request merged**